### PR TITLE
Scaffold UX tweaks

### PR DESCRIPTION
**Pull request by abaevbog**
- show an alert if one clicks on 'Test Regex', 'Run detect*', 'Run do*', 'Create Web Test' buttons without having a webpage loaded in the browser tab to guide first-time users
- add an overlay appearing on top of monaco editors in Code and Tests tabs with instructions on how to use Scaffold
- add a "Load URL" button next to the URL field. It is more discoverable than just relying on Enter keypress.
- display a spinner while the webpage is loading instead of the "Load URL" button to make it clear that loading is in progress in case it takes a while.
- add a default example.com regex as target on new translators, so it is clearer what is expected in most cases (e.g. start with a `^`, escaped dots, etc.)

https://github.com/user-attachments/assets/d29e9a6d-960f-4167-b7d9-296933d01957

@AbeJellinek, I wanted to see what some of my suggestions about Scaffold workflow would look like in practice, so I put together a few of those tweaks here. I think these would help first-time users quite a bit. The wording of instructions in the overlays and alerts may need to be tweaked a bit. What do you think about these suggestions?

### CODE REVIEW

File: chrome/content/scaffold/scaffold.js [hunk 1]
```diff
@@ -143,14 +158,15 @@ var Scaffold = new function () {
 				_editors.importGlobal = monaco;
 				_editors.import = editor;
 			}),
-			codeWin.loadMonaco({ language: 'javascript' }).then(({ monaco, editor }) => {
+			codeWin.loadMonaco({ language: 'javascript', emptyOverlayMessage: Zotero.getString('scaffold-empty-editor-code') }).then(({ monaco, editor }) => {
```
Comment (AbeJellinek) [inline chrome/content/scaffold/scaffold.js]: We don't really need to add `scaffold.ftl` for this - we generally just hardcode English strings for Scaffold. US English is the only supported locale for translator dev. (Which is probably something we should mention in the docs.)

**Review by AbeJellinek — CHANGES_REQUESTED** [review]
Thanks, a few comments but this looks good!

File: chrome/content/scaffold/scaffold.xhtml [hunk 2]
```diff
@@ -513,7 +514,8 @@
 							<hbox id="browser-url-container" align="center">
 								<label control="browser-url" value="&scaffold.tabUrl.label;"/>
 								<html:input id="browser-url" style="flex: 1;"/>
-								<button observes="validate-tests" label="&scaffold.testing.create.web;" tooltiptext="Create a new test from the current page" oncommand="Scaffold.saveTestFromCurrent('web')"/>
+								<button id="load-url-button" data-l10n-id="scaffold-browser-load-url" oncommand="Scaffold.loadURLInBrowser()"/>
```
[the remark this answers is not in the document]
Comment (abaevbog) [reply chrome/content/scaffold/scaffold.xhtml]: How does this look? I made the button a bit smaller, otherwise there was seemingly too much empty space around the arrow.

https://github.com/user-attachments/assets/04bccca8-fe27-4b87-9b4f-26db0fe83c71

File: chrome/content/scaffold/monaco/monaco.html [hunk 3]
```diff
@@ -64,6 +77,30 @@
 				
 				editor.getModel().setEOL(monaco.editor.EndOfLineSequence.LF);
 
+
+				// Display overlay message when editor is empty, if specified
+				if (params.emptyOverlayMessage) {
+					let overlayNode = document.createElement('div');
+					overlayNode.className = 'empty-editor-overlay';
+
+					overlayNode.innerHTML = params.emptyOverlayMessage;
```
[the remark this answers is not in the document]
Comment (abaevbog) [reply chrome/content/scaffold/monaco/monaco.html]: Right, `innerText` is cleaner, I agree.

File: chrome/content/scaffold/scaffold.js [hunk 4]
```diff
@@ -1070,6 +1097,36 @@ var Scaffold = new function () {
...
+	};
+
+	/**
+	 * Check if there is a webpage loaded in the browser tab when translator type is 'web'.
+	 * If not, alert the user and navigate to the browser tab.
+	 * @returns {boolean} True if a webpage is loaded for web translators, false otherwise.
+	 */
+	this.ensureWebpageLoadedIfNeeded = function () {
+		if (!_browserLoadedURL && document.getElementById('checkbox-web').checked) {
+			Services.prompt.alert(null, " A loaded webpage is required.", "Load a relevant webpage in the Browser tab to use this feature for the web translator.");
```
Comment (AbeJellinek) [inline chrome/content/scaffold/scaffold.js]: Maybe this wording?

```suggestion
			Services.prompt.alert(null, "Webpage Not Loaded", "You need to load a webpage in the Browser tab first.");
```

**Comment by AbeJellinek** [comment]
But that'll basically make the alert never show, right?

Maybe we could show the alert instead of logging in the final `else` block if it's a web translator (`document.getElementById('checkbox-web').checked`).

**Comment by abaevbog** [comment]
Yes, sorry, I realized it would not work right after pushing. I think checking in the final `else` block for if we should show the alert or not works the best.

Also, with the current setup of selecting `web` vs `search` commands to run, I realized that we needed to bring back the ability to clear the URL from the browser (which I had removed thinking that if the url i empty, we should just do nothing).

**Comment by abaevbog** [comment]
Just for added control, the logic here is relying on the `_browserLoadedURL` variable that we are setting in the brower's progress listener. That way, we don't count a browser as having a loaded URL if, for example, one mistypes a link and nothing is actually loaded.

**Comment by abaevbog** [comment]
> Just for added control, the logic here is relying on the _browserLoadedURL variable that we are setting in the brower's progress listener. 

Actually, with some more testing, I feel like this just lead to more edge cases... Like if the web page loads more data after initial load, _browserLoadedURL would flip from positive to negative and the alert might theoretically appear when the page is technically loaded. Since the main end goal here is just usability and having an indication spinner that the page is loading, it's probably best to just check against `_browser.currentURI.spec` like before.

**Comment by AbeJellinek** [comment]
OK, that looks good!

Last thing I'm not sure about here is moving the Create Web Test button to Tests. The Import/Search equivalents are still under another tab, so that just feels confusing. Could we move it back to Browser (to the next row, under the URL field and Load URL button), but add an explanatory label under Tests? Maybe something like "Create tests from the Browser and Test Input tabs."

**Comment by abaevbog** [comment]
> Last thing I'm not sure about here is moving the Create Web Test button to Tests. The Import/Search equivalents are still under another tab, so that just feels confusing.

Yeah, that makes a lot of sense. I have been focusing on just Web translators but in the context of other translator types, it's just weird to move only one button creating tests out of 3. If we were to place it under the URL field, I am afraid there might be too much empty space between browser and the input field... What do you think about just placing it after the arrow icon? I feel like there's enough space to not look too crowded?

Screenshot 2025-11-06 at 1 30 53 PM

And then this is what the empty message in the tests editor looks like. My main idea behind moving "Create Web Test" button was to avoid one looking at the empty code editor without having any idea what to do here. Now that we have a small message, I think that problem is very much addressed.

Screenshot 2025-11-06 at 1 30 45 PM

Speaking of the empty code editors, I didn't add anything into the final Test Input tab yet, since I didn't run into it during my first translator experiment. I also haven't been able to find a good reference point as to what should go there... Can you think of what would be an appropriate for that editor? If not, I think it's not a huge deal now, since the majority of translators are web.

File: chrome/content/scaffold/scaffold.js [hunk 5]
```diff
@@ -40,6 +40,13 @@ ChromeUtils.defineLazyGetter(lazy, 'shellPathPromise', () => {
 		.then(s => s.trimEnd());
 });
 
+// Text to display in monaco editors when they are empty
+const CODE_TAB_INFO = `1. Use Tools > Insert Template menu option or the "+" icon above to insert a template for the translator.
+2. Use Run detect* and Run do* buttons in the toolbar to run your translator on the URL loaded in the browser tab.
+3. For reference, use Help > List All Item Types to list all available Zotero item types that your translator can create.
+4. For reference, use Help > List all Fields for Item Type to log the fields of the Zotero item(s) your translator is creating.`;
```
Comment (AbeJellinek) [inline chrome/content/scaffold/scaffold.js]:
```suggestion
3. Use Help → List All Item Types and Help → List All Fields for Item Type for information on the item types and fields your translator can return.
```

File: chrome/content/scaffold/monaco/monaco.html [hunk 6]
```diff
@@ -13,6 +13,18 @@
...
+		.monaco-editor .empty-editor-overlay {
+			position: absolute;
+			top: 0;
+			left: 0;
+			right: 0;
+			padding: 0 64px;
+			color: #ffffff8c; /* dark fill-secondary */
+			font-size: 12px;
+			line-height: 2;
+			pointer-events: auto;
```
[the remark this answers is not in the document]
Comment (abaevbog) [reply chrome/content/scaffold/monaco/monaco.html]: You are right, pointer events are indeed not necessary here. I believe I was thinking about accounting for clickable elements that we might want in the placeholder, like links. But we don't have any of those, so, yeah, I changed it for `pointer-events: none` here.

**Comment by abaevbog** [comment]
Ok, I pushed all of the remaining small edits above. Thank you for going over this (and the proof-reading of editor placeholders)!

**Comment by AbeJellinek** [comment]
Thank you so much!

**Comment by abaevbog** [comment]
Great, hopefully that'll help to clarify things a bit for first time users! I'll go and update that writeup about writing the first translator to have screenshots of the new UI
